### DISCUSSION TITLE
[v4] Ampersand Alias not working as expected

### DISCUSSION THREAD

**Discussion by moshie** [tailwindlabs/tailwindcss#15165] [Help]
**What version of Tailwind CSS are you using?**
```
@tailwindcss/postcss: "4.0.0-beta.2"
tailwindcss: "4.0.0-beta.2"
```

**What build tool (or framework if it abstracts the build tool) are you using?**

NextJS: 15.0.3

**What version of Node.js are you using?**

v20

**What browser are you using?**

Arc

**What operating system are you using?**

macOS

**Describe your issue**

When writing css such as:

```css
.btn {
  @apply px-4 py-2 rounded;

  &-primary {
    @apply bg-blue-500 text-white;
  }
}
```

Then the `.btn-primary` class is not generated and present in the css file.

**Comment by philipp-spiess** [comment]
Hey there! I don't think a `.btn-primary` class is supposed to be created, e.g. check out this example directly in the browser without any Tailwind post-processing: https://play.tailwindcss.com/FdeyotGGC9?file=css

What you refer to is probably a feature of another CSS pre-processor (maybe [SCSS](https://sass-lang.com/playground/#eJwzNNRLKslTqOZSUEhKTM5OL8ovzUvRTc7PyS+yUihKTbHmAsqo6RYUZeYmFlWC1WFTmZRTmmoNlKvlquUCAEMfGcc=)?). If you want this to work, you'll have to ensure that you run the according pre-processors before Tailwind in your postcss config.

In general though, Tailwind v4 uses browser-based CSS nesting and [lightningcss](https://lightningcss.dev/playground/index.html#%7B%22minify%22%3Atrue%2C%22customMedia%22%3Atrue%2C%22cssModules%22%3Afalse%2C%22analyzeDependencies%22%3Afalse%2C%22targets%22%3A%7B%22chrome%22%3A6225920%7D%2C%22include%22%3A0%2C%22exclude%22%3A0%2C%22source%22%3A%22.btn%20%7B%5Cn%20%20color%3Ared%3B%5Cn%5Cn%20%20%26-primary%20%7B%5Cn%20%20%20%20color%3Ablue%3B%5Cn%20%20%7D%5Cn%7D%22%2C%22visitorEnabled%22%3Afalse%2C%22visitor%22%3A%22%7B%5Cn%20%20Color(color)%20%7B%5Cn%20%20%20%20if%20(color.type%20%3D%3D%3D%20'rgb')%20%7B%5Cn%20%20%20%20%20%20color.g%20%3D%200%3B%5Cn%20%20%20%20%20%20return%20color%3B%5Cn%20%20%20%20%7D%5Cn%20%20%7D%5Cn%7D%22%2C%22unusedSymbols%22%3A%5B%5D%2C%22version%22%3A%22local%22%7D) to flatten the nesting.

**Reply by moshie** [reply]
Hey 👋 thanks for getting back to me! 

I might be mistaken but didn't this work before with the `tailwindcss/nesting` postcss plugin which wraps https://github.com/postcss/postcss-nested ?

I guess not implementing this will introduce breaking changes when migrating from v3?

**Reply by philipp-spiess** [reply]
Hmm, I see. I didn't realize this postcss plugin does non-standard things like this. We should probably update our upgrade tool to fix this, then. Let me bring this up with the team 👍 

In your case, if you depend on this behavior, it might be best if you do not remove `tailwindcss/nesting` from your postcss config yet and still keep using it if you have tons of usages of this pattern and can't migrate right away.

**Reply by moshie** [reply]
@philipp-spiess Hey any updates?

**Reply by philipp-spiess** [reply]
@moshie No updates I'm afraid, I recommend migrating away from this non-standards pattern.

**Reply by mortscode** [reply]
@philipp-spiess The ampersand is standard in modern browsers, correct? https://developer.mozilla.org/en-US/docs/Web/CSS/Nesting_selector

This is a bummer for anyone taught BEM. I have a lot of re-factoring to do on this project. Yikes.

**Reply by philipp-spiess** [reply]
@mortscode ampersand is standard, correct, but the `&-primary` syntax to concatenate class names is not. If you rely on this, you can just continue to use `postcss-nested` though!

**Comment by yakshcore** [comment]
<h2>✅ The Fixes</h2>
<h3>✅ Option 1: Flatten the Nesting</h3>
<p>The cleanest fix is: <strong>don’t nest</strong> when you’re using Tailwind classes + <code inline="">@apply</code>. Write it out like this:</p>
<pre><code class="language-css">.btn {
  @apply px-4 py-2 rounded;
}

.btn-primary {
  @apply bg-blue-500 text-white;
}
</code></pre>
<p>This way, Tailwind clearly sees the <code inline="">.btn-primary</code> class, and it won’t get purged. ✂️</p>
<hr>
<h3>✅ Option 2: Safelist It (if you really wanna nest)</h3>
<p>If you <em>must</em> use nesting (<code inline="">&amp;-primary</code>) and can’t flatten your code, you need to <strong>safelist</strong> the generated class manually in <code inline="">tailwind.config.js</code>:</p>
<pre><code class="language-js">module.exports = {
  content: ['./src/**/*.{js,ts,jsx,tsx,html}'],
  safelist: [
    'btn-primary',
  ],
}
</code></pre>
<p>But honestly, this is annoying and fragile if you’re gonna do it for 20+ nested classes.</p>
<hr>
<h3>🛠️ Extra: If you're using <code inline="">postcss-nesting</code></h3>
<p>Make sure your <code inline="">postcss.config.js</code> has this in the correct order:</p>
<pre><code class="language-js">module.exports = {
  plugins: {
    'postcss-nesting': {},
    tailwindcss: {},
    autoprefixer: {},
  },
}
</code></pre>
<p>Also ensure that you’re using <strong>PostCSS nesting plugin</strong>, not <code inline="">postcss-nested</code>, because that’s deprecated and behaves differently.</p>
